Switch horses when tweaking combat mods

MatthewDarling opened the pull request:
## Description of changes

`sl_ascend`  grabs the Dark Horse before each adventure for its -combat modifier. Upon reaching the war in HC DG, I found it wouldn't do the Lighthouse quest because it couldn't get enough +combat to override the Dark Horse. It would just say:

`Something is keeping us from getting a suitable combat rate, we have: -5.0 and Lobsterfrogmen.`

While KoLMafia's maximizer does know about Horsery, my understanding is that we don't want to call into it regularly. So, I've added manual checks for Horsery settings in the combat modifier functions.

Similar to the Asdon, it costs meat to switch, so I put it towards the bottom of the list. I put it before the Asdon check, assuming Asdon costs more than 500 to switch. At any rate, we want the cheapest one to be first, so I'll switch if needed.

With these changes, `sl_ascend` did the Lighthouse quest, so I'm happy :)

## Stuff to maybe look at fixing

1. Check if Horsery is unrestricted. I've seen the `$item` syntax for checking this, but Horsery is a location, so I'm not sure how to check that. Right now we're using `_horseryAvailable`, and I don't know if that checks for restrictions on usage

jaspercb commented on RELEASE/scripts/sl_ascend/sl_util.ash:
```diff
@@ -2769,6 +2769,14 @@ boolean providePlusCombat(int amt, boolean doEquips)
+           && (get_property("_horsery") == "dark horse")
+           && (my_meat() >= 1000))
+        {
+                cli_execute("horsery normal");
```
use `getHorse(string type)` from sl_mr2017.ash

jaspercb commented on RELEASE/scripts/sl_ascend/sl_util.ash:
```diff
@@ -2834,6 +2842,14 @@ boolean providePlusNonCombat(int amt, boolean doEquips)
+           && (get_property("_horsery") != "dark horse")
+           && (my_meat() >= 1000))
+        {
+                cli_execute("horsery dark");
```
I don't understand how this doesn't contend with the horsery setting LX_craftAcquireItems() until you run out of meat.
```
if(sl_my_path() != "Community Service")
	{
		getHorse("noncombat");
```

Maybe use something parallel to the Dark Gyffte form solution? (instead of changing the horse immediately, set a local preference, check in the preadventure or in ccAdv whether we're in the desired state, if so fix it if possible).

MatthewDarling replied:
I was watching to see if my meat would disappear in the span of a few seconds after putting this in, haha. At the very least, I can confirm it didn't spend all my meat, cause I have more than I started the day with.

But it's probably spending more than necessary. Your alternate approach will probably fix that, so I'll make that switch rather than dig through my session logs.

jaspercb replied:
I *think* it's spending 1K meat a turn (the cost of two switches), but I'm not sure of that.

jaspercb commented:
My suspicion is that sausage grinding is broken because you have no meat from the aforementioned tug-of-war taking place.

MatthewDarling commented:
@jaspercb Thanks for the feedback! Tried to address it, let me know how it looks.

Note: **Not actually tested**. The initial buggy state of the PR allowed me to complete today's ascension, and I didn't want to risk my day 1 meat of the new ascension. When I hit Sonofa Beach in a couple days, I'll give it a shot :)

If there are any issues, I probably won't be able to fix them until ~Tuesday next week.

MatthewDarling commented:
Tested today: still not working :( The [Lobsterfrogmen check](https://github.com/soolar/sl_ascend/blob/master/RELEASE/scripts/sl_ascend.ash#L7418) doesn't use `providePlusCombat`.

I've made a bunch of tweaks locally, but I'm still not getting ideal behaviour. I've got it so that `L12_sonofaPrefix()` will return the Dark Horse, but on the next adventure, `LX_craftAcquireItems()` rents the horse before the L12 quest code can run.

Is there something I missed about the DG form handling that avoids this issue? It seems like I need some way of saying "okay I want this setting for a certain duration of time, don't allow any changes, okay I'm done go back to default".

soolar commented:
I'm not entirely sure what's wrong with it, I'll look more in depth tomorrow.
Also, minor thing, could you make getHorse set sl_desiredHorse instead of relying on calling both together?

MatthewDarling commented:
Pushed the dumb commits I had locally just to illustrate what's happening. And also cause I wanted to rebase onto `master` so I can test a little :)

I'm thinking maybe a "set desired only if there's no existing value" function might fix things. Using that in `LX_craftAcquireItems()` would mean Dark Horse remains the fallback value. And I guess clear the desired value in a post-adventure task? Then each adventure... should... recompute the optimal horse and only change if needed.

----------

Re: the various called to `setHorse`, yep I can address that. Reviewed the Vampyre form code again and realized that the pre-adventure is more of a backup plan in case something happens between setting desired + starting adventure. I'd been confused by the immediate `switchForm` calls in the functions that set the desired form, but it makes sense now.

jaspercb commented:
The pre-adventure isn't *just* a backup plan, it clears away our current form when we're in a form but want to be in no form. Other than that, correct.

MatthewDarling commented:
Latest two commits made `getHorse()` apply the `sl_desiredHorse` setting, and switched the combat modifier functions to use that to immediately make changes.

This has improved things a lot - `L12_sonofaPreix()` was already using `providePlusCombat()`, so now it returns the horse and tries to do the Lighthouse quest. And it isn't wasting 1000 meat per adventure :D 

...Instead, it's wasting 250 meat per adventure. It alternates between returning the horse and getting dark horse. Is it possible the preadventure code is being run twice before an actual adventure gets spent? Presumably after `LX_acquireCraftItems()` but before `L12_sonofaPrefix()`.

Here's a filtered log:

```
> Turn(581): Starting with 28 left at Level: 13
> Encounter: -5.0   Exp Bonus: 12.75
> Delay between adventures... beep boop... 

use 1 Gathered Meat-Clip
You gain 565 Meat

Visiting The Horsery
Took choice 1266/5: unknown
choice.php?pwd&whichchoice=1266&option=5
Returned your horse
> Starting preadventure script...

cast 1 Flock of Bats Form
You lose an intrinsic: Bats Form

Visiting The Horsery
Took choice 1266/5: unknown
choice.php?pwd&whichchoice=1266&option=5
Returned your horse
> Pre Adventure at Sonofa Beach done, beep.

[582] Sonofa Beach
Encounter: We Apologize To Our English Players For the Lameness of the Pun
You acquire bottle of rum (3)
You acquire lemon (3)
You acquire lime (3)
> Post Adventure done, beep.
> Turn(582): Starting with 27 left at Level: 13
> Encounter: 0.0   Exp Bonus: 14.837500000000002
> Delay between adventures... beep boop... 

equip weapon oversized pipe
> Starting preadventure script...

Visiting The Horsery
Took choice 1266/2: Rent This Horse
choice.php?pwd&whichchoice=1266&option=2
Chose the dark horse
> Pre Adventure at Sonofa Beach done, beep.

[583] Sonofa Beach
Encounter: We Apologize To Our English Players For the Lameness of the Pun
You acquire bottle of rum (3)
You acquire lemon (3)
You acquire lime (3)
> Post Adventure done, beep.
> Turn(583): Starting with 26 left at Level: 13
> Encounter: -5.0   Exp Bonus: 14.837500000000002
> Delay between adventures... beep boop... 

Visiting The Horsery
Took choice 1266/5: unknown
choice.php?pwd&whichchoice=1266&option=5
Returned your horse
> Starting preadventure script...

Visiting The Horsery
Took choice 1266/5: unknown
choice.php?pwd&whichchoice=1266&option=5
Returned your horse
> Pre Adventure at Sonofa Beach done, beep.

[584] Sonofa Beach
Encounter: Hippies Don't Surf
You gain 30 Cheek
> Post Adventure done, beep.
> Turn(584): Starting with 25 left at Level: 13
> Encounter: 0.0   Exp Bonus: 14.837500000000002
> Delay between adventures... beep boop... 
> I think I'm paranoid and complicated
> I think I'm paranoid, manipulated
> Starting preadventure script...

Visiting The Horsery
Took choice 1266/2: Rent This Horse
choice.php?pwd&whichchoice=1266&option=2
Chose the dark horse
> Pre Adventure at Sonofa Beach done, beep.

[585] Sonofa Beach
Encounter: Mystical Crap
You gain 33 Magicalness
> Post Adventure done, beep.
> Turn(585): Starting with 24 left at Level: 13
> Encounter: -5.0   Exp Bonus: 14.837500000000002
> Delay between adventures... beep boop... 

Visiting The Horsery
Took choice 1266/5: unknown
choice.php?pwd&whichchoice=1266&option=5
Returned your horse
> Starting preadventure script...

Visiting The Horsery
Took choice 1266/5: unknown
choice.php?pwd&whichchoice=1266&option=5
Returned your horse
> Pre Adventure at Sonofa Beach done, beep.

[586] Sonofa Beach
Encounter: Save the Texas Prairie Chicken
You acquire an item: wool hat
> Post Adventure done, beep.
> Turn(586): Starting with 23 left at Level: 13
> Encounter: 0.0   Exp Bonus: 14.837500000000002
> Delay between adventures... beep boop... 
> Starting preadventure script...

Visiting The Horsery
Took choice 1266/2: Rent This Horse
choice.php?pwd&whichchoice=1266&option=2
Chose the dark horse
> Pre Adventure at Sonofa Beach done, beep.

[587] Sonofa Beach
Encounter: What a Maroon
You acquire an item: goatskin umbrella
> Post Adventure done, beep.
> Turn(587): Starting with 22 left at Level: 13
> Encounter: -5.0   Exp Bonus: 14.837500000000002
```

(I added `print(get_stack_trace())` into `horsePreAdventure()` but the output was `aggregate {string file; string name; int line;} [5]`; if I can print the full structure that would help debug)

jaspercb commented:
Be a pragmatist, move the `horseDark();` call somewhere after `LX_craftAcquireItems();`.

MatthewDarling commented:
Fiiiiiinally figured out the issue. [Second-to-latest commit](https://github.com/soolar/sl_ascend/pull/122/commits/c58c56d1c6b1ab7465ee44f09c860e51d7766b96) has some details. Basically, I was mishandling horse state: `providePlusCombat` was looking at no-horse state as "we've got the +combat that we need, leave horse alone" which led to the default dark horse getting picked.

I added a function `horseNormalize` to make `horseMaintain` simpler. But `getHorse` could use it, too  - thoughts?

MatthewDarling commented on RELEASE/scripts/sl_ascend/sl_mr2017.ash:
```diff
@@ -1433,6 +1433,48 @@ int horseCost()
+		case "":	   return "return";
+	}
+
+	if (contains_text(horseText, "normal horse"))
```
I had put these here because `getHorse` has it in some cases. But it seems like the literal value "dark horse" is always used in `_horsery`, and maybe the others are too. In which case these could go into the `switch`, instead.

Need to test it with all the different horse types first, though.

soolar commented:
Please do make all of the horse related functions that could use it use `horseNormalize`, seems like the best approach. Also, why is `sl_desiredHorse` expected to be `regen`, `noncombat`, `random`, or `resistance` instead of the same values as `_horsery` uses, since that's what horseNormalize returns?

Other than those adjustments, looks like this is pretty much ready to merge! Thanks! 👍
